src/storage/layout.cpp

```diff
@@ -0,0 +1,354 @@
+//
+// layout.cpp
+//
+// Identification: src/storage/layout.cpp
+//
+// Copyright (c) 2015-2018, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include <sstream>
+#include <string>
+
+#include "catalog/column.h"
+#include "catalog/schema.h"
+#include "storage/layout.h"
+#include "util/stringbox_util.h"
+
+namespace peloton {
+namespace storage {
+
+// Constructor for the layout class with column_count
+// The default layout is always a ROW_STORE
+Layout::Layout(const oid_t num_columns, LayoutType layout_type)
+    : num_columns_(num_columns), layout_type_(layout_type) {}
+
+// Constructor for the Layout class with column_map
+Layout::Layout(const column_map_type &column_map)
+    : num_columns_(column_map.size()), column_layout_(column_map) {
+  // Figure out the layout type if not present
+  bool row_layout = true, column_layout = true;
+  for (oid_t column_id = 0; column_id < num_columns_; column_id++) {
+    if (column_layout_[column_id].first != 0) {
+      row_layout = false;
+    }
+    if (column_layout_[column_id].first != column_id) {
+      column_layout = false;
+    }
+  }
+
+  // Assign the layout_type_
+  // If a table has one column, it would be LayoutType::Row
+  if (row_layout) {
+    layout_type_ = LayoutType::ROW;
+    layout_oid_ = ROW_STORE_OID;
+  } else if (column_layout) {
+    layout_type_ = LayoutType::COLUMN;
+    layout_oid_ = COLUMN_STORE_OID;
+  } else {
+    // layout_oid_ is set to INVALID_OID, indicating that this
+    // layout is not stored in the catalog and thus not persistent.
+    // To be used only in TempTable or Tests.
+    layout_type_ = LayoutType::HYBRID;
+    layout_oid_ = INVALID_OID;
+  }
+
+  if (layout_type_ != LayoutType::HYBRID) {
+    // Since this is a predefined layout, we don't need the map
+    column_layout_.clear();
+  }
+}
+
+// Constructor for Layout class with predefined layout_oid
+Layout::Layout(const column_map_type &column_map, oid_t layout_id)
+    : layout_oid_(layout_id),
+      num_columns_(column_map.size()),
+      column_layout_(column_map) {
+  if (layout_oid_ == ROW_STORE_OID) {
+    layout_type_ = LayoutType::ROW;
+  } else if (layout_oid_ == COLUMN_STORE_OID) {
+    layout_type_ = LayoutType::COLUMN;
+  } else {
+    layout_type_ = LayoutType::HYBRID;
+  }
+
+  if (layout_type_ != LayoutType::HYBRID) {
+    // Since this is a predefined layout, we don't need the map
+    column_layout_.clear();
+  }
+}
+
+// Sets the tile id and column id w.r.t that tile corresponding to
+// the specified tile group column id.
+void Layout::LocateTileAndColumn(oid_t column_offset, oid_t &tile_offset,
+                                 oid_t &tile_column_offset) const {
+  // Ensure that the column_offset is not out of bound
+  PELOTON_ASSERT(num_columns_ > column_offset);
+
+  // For row store layout, tile id is always 0 and the tile
+  // column_id and tile column_id is the same.
```

Comment:
Comment text is unclear...
and the tile column_id and tile column_id is the same